openwisp-pre-push-hook

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
+
+import argparse
+import json
+import os
+import subprocess
+import sys
+
+
+def get_top_level_dir():
+    top_level_dir = subprocess.Popen(
+        ['git', 'rev-parse', '--show-toplevel'], stdout=subprocess.PIPE
+    ).communicate()[0].decode('utf-8').strip()
+    return top_level_dir
+
+def make_executable(f):
+    task = subprocess.Popen(
+        ['chmod', '+x', f], stdout=subprocess.PIPE, stderr=subprocess.PIPE
+    )
+    err = task.communicate()[1]
+    return err
+
+def run_checks():
+    try:
+        import openwisp_utils
+    except:
+        raise ImportError(
+            "'openwisp_utils' failed to import. Make sure all dependencies "
+            "are installed and virtual environment is activated."
+        )
+    checks = ['run-qa-checks', 'runtests.py', 'runtests']
```

Comment:
here's some more problems, we need to give this a bit more thought if we want it to really be useful.

For some repositories, running all the tests can take a long time, so we usually run tests using --parallel.
But in openwisp-monitoring we can't use `--parallel`.

More over, in some repos we also run tests with `SAMPLE_APP=1`.

Dealing with these cases here would complicate things too much.

I think we could just delegate everything to the `run-qa-checks` script in every repository: this script would also run tests, eg:

- `./run-qa-checks` runs `./runtests.py --parallel` and `SAMPLE_APP=1 ./runtests.py --parallel` in openwisp-radius
- `./run-qa-checks` runs `./runtests.py` and `SAMPLE_APP=1 ./runtests.py` in openwisp-monitoring
- `./run-qa-checks` runs `yarn test` in openwisp-wifi-login-pages

But we should allow each `./run-qa-checks` script to be called from the CI builds, eg:

- `./run-qa-checks --ci`, so we can skip the `--parallel` flag in some cases otherwise the test coverage would drop.

@atb00ker @pandafy @nepython @NoumbissiValere @okraits what do you think? Any other useful suggestion?

Reply:
This looks good, @Purhan I would suggest before this is merged, open a related PR in another module like controller and use it there for actual testing! :smile:

Reply:
> ./run-qa-checks --ci, so we can skip the --parallel flag in some cases otherwise the test coverage would drop.

What is we instead, use a flag for the hook (`--hook`) and everything else works as normal with minimal changes?